stor/dx.py

```diff
@@ -0,0 +1,513 @@
+                                                      entry['describe']['folder'])
+                                        / entry['describe']['name'])
+        return ans_list
+
+    def listdir_iter(self, canonicalize=False):
+        """Iterate the path as a dir, returning top-level directories and files.
+
+        Args:
+            canonicalize (boolean): whether to return canonicalized paths
+
+        Returns:
+            Iter[DXPath]: Iterates over listed files directly within the resource
+        """
+        folders = self.listdir(only='folders', canonicalize=canonicalize)
+        for folder in folders:
+            yield folder
+        for data in self.walkfiles(
+                        canonicalize=canonicalize,
+                        recurse=False
+                    ):
+            yield data
+
+    def walkfiles(self,
+                  pattern=None,
+                  canonicalize=False,
+                  recurse=True,
+                  starts_with=None,
+                  limit=None,
+                  category=None):
+        """Iterates over listed files that match an optional pattern.
+
+        Args:
+            pattern (str): glob pattern to match the filenames against.
+            canonicalize (boolean): whether to return canonicalized paths
+            recurse (boolean): whether to look in subfolders of folder as well
+            starts_with (str): Allows for an additional search path to
+                be appended to the resource of the dx path. Note that this
+                resource path is treated as a directory
+            limit (int): Limit the amount of results returned
+            category (str): Restricting class : One of 'record', 'file', 'gtable,
+                'applet', 'workflow'
+
+        Returns:
+             Iter[DXPath]: Iterates over listed files that match an optional pattern.
+        """
+        proj_id = self.canonical_project
+        proj_name = self.virtual_project
+        kwargs = {
+            'project': proj_id,
+            'name': pattern,
+            'name_mode': 'glob',
+            # the query performance is similar w/wo describe field,
+            # hence no need to customize query based on canonicalize flag
+            'describe': {'fields': {'name': True, 'folder': True}},
+            'recurse' : recurse,
+            'classname': category,
+            'limit': limit,
+            'folder': (self.resource or '/') + (starts_with or '')
+        }
+        if self.resource and not self._is_folder():  # if path is a file path
+            yield []
+            return
+        list_gen = dxpy.find_data_objects(**kwargs)
+        for obj in list_gen:
+            if canonicalize:
+                yield DXCanonicalPath('dx://{}:/{}'.format(obj['project'], obj['id']))
+            else:
+                dx_p = DXVirtualPath(self.drive + proj_name + ':' + obj['describe']['folder'])
```

Comment:
I think this'd be clearer written as:

```
'{drive}{proj_name}:{folder}/{name}'.format(drive=drive, proj_name=proj_name, folder=obj['describe']['folder'], name=obj['describe']['name'])
``` which I believe does the same thing but avoids a level of indirection

Reply:
Yep, changed here and other places where it would be more intuitive to write like this.